client/src/main/java/io/pravega/client/batch/impl/BatchClientFactoryImpl.java

```diff
@@ -174,4 +216,106 @@ public List<SegmentRange> getSegmentRangeBetweenStreamCuts(final StreamCut start
+        return offsetLocated.getOffset();
+    }
+
+    private boolean checkIfNextOffsetSame(long existingOffset, long nextOffset) {
+        return existingOffset == nextOffset;
+    }
+
+    private void checkSuccessorSegmentOffset(Map<Segment, Long> nextPositionsMap, Map<Segment, Long> scaledSegmentsMap, long approxDistanceToNextOffset) {
+        log.debug("checkSuccessorSegmentOffset() -> Segments that may have scaled = {}", scaledSegmentsMap);
+        scaledSegmentsMap.entrySet().stream().forEach(entry -> {
+            CompletableFuture<StreamSegmentsWithPredecessors> getSuccessors = controller.getSuccessors(entry.getKey());
+            Map<SegmentWithRange, List<Long>> segmentToPredecessorMap = getSuccessors.join().getSegmentToPredecessor();
+            int size = segmentToPredecessorMap.size();
+            if (size > 1) { //scale up happened to the segment
+                for (SegmentWithRange segmentWithRange : segmentToPredecessorMap.keySet()) {
+                    Segment segment = segmentWithRange.getSegment();
+                    long nextOffset = getNextOffsetForSegment(segment, approxDistanceToNextOffset);
+                    nextPositionsMap.put(segment, nextOffset);
+                }
+            } else if (size == 1) { //scale down happened to the segments
+                List<Long> segmentIds = nextPositionsMap.keySet().stream().map(x -> x.getSegmentId()).collect(Collectors.toList());
+                // Check for any of the predecessor which is present in nextPositionsMap. If so, we will not proceed to successor segment
+                boolean isJoint  = segmentToPredecessorMap.values().stream().findFirst().get().stream().anyMatch(segmentIds::contains);
+                if (!isJoint) {
+                    Long segmentId = segmentToPredecessorMap.keySet().stream().findFirst().get().getSegment().getSegmentId();
+                    if (!segmentIds.contains(segmentId)) {
+                        Segment segment = segmentToPredecessorMap.keySet().stream().findFirst().get().getSegment();
+                        long nextOffset = getNextOffsetForSegment(segment, approxDistanceToNextOffset);
+                        nextPositionsMap.put(segment, nextOffset);
+                    }
+                } else {
+                    nextPositionsMap.put(entry.getKey(), entry.getValue());
+                }
+            } else { //Segment is neither sealed and nor scaled
+                nextPositionsMap.put(entry.getKey(), entry.getValue());
+            }
+
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @SneakyThrows({ConnectionFailedException.class})
+    private <T extends Reply> T transformReply(Reply reply, Class<T> klass) {
+        if (klass.isAssignableFrom(reply.getClass())) {
+            return (T) reply;
+        }
+        closeConnection(reply);
+        if (reply instanceof WireCommands.NoSuchSegment) {
```

Comment:
I think we need to handle **SearchFailedException** as well or need to through it to user.

Reply:
Added searchfailed exception.